### PR TITLE
encoding: Replace float and decimal encodings with faster approaches

### LINKED ISSUE #5192: util/encoding: consider changing float encoding
https://github.com/google/orderedcode/blob/master/orderedcode.go#L246-L257 shows that we can encode the float64 bits reinterpreted as an int64. I think we'll have to do some additional handling of NaN, positive and negative infinity, but that shouldn't be a problem.

### PR DESCRIPTION
**Pull request by nvb**
## Float Encoding

Fixes #5192.

This approach was inspired by [orderedcode.go](https://github.com/google/orderedcode/blob/master/orderedcode.go#L246). It reinterprets the bits
of a float64 as an int64, and messes with this reinterpretation to get
the ordering correct. For positive floats, the encoding of the 64 bits
(in IEEE 754 format) is directly encoded using `EncodeUint64Ascending`.
For negative floats, we keep the sign bit and invert all other bits,
encoding this value using `EncodeUint64Descending`.

```
name           old time/op    new time/op    delta
EncodeFloat-4     314ns ± 7%      23ns ± 3%  -92.53%         (p=0.000 n=10+9)
DecodeFloat-4     308ns ±12%      30ns ± 1%  -90.24%         (p=0.000 n=10+9)

name           old alloc/op   new alloc/op   delta
EncodeFloat-4    0.00B ±NaN%    0.00B ±NaN%     ~     (all samples are equal)
DecodeFloat-4    0.00B ±NaN%    0.00B ±NaN%     ~     (all samples are equal)

name           old allocs/op  new allocs/op  delta
EncodeFloat-4     0.00 ±NaN%     0.00 ±NaN%     ~     (all samples are equal)
DecodeFloat-4     0.00 ±NaN%     0.00 ±NaN%     ~     (all samples are equal)
```
## Decimal Encoding

Now that we are no longer tied to the sqlLite float encoding for
decimals, we can go nuts with a custom encoding. This commit proposes a
new encoding that uses the bytes directly from the underlying `big.Int`
to represent the decimal's mantissa. After a lot of tinkering the
approach reduced decimal encoding to zero allocations! It uses a
couple of tricks to achieve that, including a binary digit lookup table
(because we absolutely **do not** want to have to string format the decimal!)
and a few routines working on slices of little-endian `big.Words` so that
we can flip the bytes into the encoding key directly. Decoding was also
improved by reducing allocations in a few key spots. These included
using the binary digit lookup table, performing ones-complements on the
keys themselves, and better handling the decimal's underlying `big.Int`.

```
name             old time/op    new time/op    delta
EncodeDecimal-4     470ns ± 2%     161ns ± 1%   -65.81%    (p=0.000 n=9+9)
DecodeDecimal-4     622ns ± 2%     239ns ± 1%   -61.66%  (p=0.000 n=10+10)

name             old alloc/op   new alloc/op   delta
EncodeDecimal-4     89.0B ± 0%     0.0B ±NaN%  -100.00%  (p=0.000 n=10+10)
DecodeDecimal-4      176B ± 0%       96B ± 0%   -45.45%  (p=0.000 n=10+10)

name             old allocs/op  new allocs/op  delta
EncodeDecimal-4      3.00 ± 0%     0.00 ±NaN%  -100.00%  (p=0.000 n=10+10)
DecodeDecimal-4      4.00 ± 0%      2.00 ± 0%   -50.00%  (p=0.000 n=10+10)
```
### Disclaimer

_I marked this as noteworthy because this will **break existing clusters**. It also 
**needs** to be resolved by beta._

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5337)

### CODE REVIEW

File: util/encoding/float.go [hunk 1]
```diff
@@ -12,57 +12,50 @@
...
 		return append(b, floatZero)
 	}
-	// TODO(nvanbenschoten) Switch to a base-256 mantissa.
-	e, m := floatMandE(f, b[len(b):])
-	return encodeMandE(b, f < 0, e, m)
+	i := int64(math.Float64bits(f))
+	if i < 0 {
+		i = math.MinInt64 - i
+		b = append(b, floatNeg)
+		return EncodeUint64Descending(b, uint64(-i))
```
Comment (petermattis) [inline util/encoding/float.go]: The subtraction, negating and inverting seems excessive here. Can't we directly call `EncodeUint64Descending(b, uint64(i))`? Which is really just a call to `EncodeUint64Ascending(b, ^uint64(i))`? I'm thinking this should look like:

```
u := math.Float64bits(f)
if u&(1<<63) != 0 {
  u = ^u
  b = append(b, floatNeg)
} else {
  b = append(b, floatPos)
}
return EncodeUint64Ascending(b, u)
```

**Comment by knz** [comment]
I'm not sure it's a great idea to merge the encoding for positive and negative zero. This break math. 

Consider for example the following basic principle: x / ceil(x) should have the same sign as x for all x. 
With signed zero you get -0.5 / ceil(-0.5) = -0.5 / -0.0 = -Inf.    If you store +/- 0 in the same way, that means that the formula loses its sign property if you store the operand in a table. Do we really want that?

**Comment by petermattis** [comment]
Reviewed 4 of 4 files at r1.
Review status: 1 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[util/encoding/decimal.go, line 32 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hDSVBsOimZfBA5kd:-KD3hDSVBsOimZfBA5ke:1197648929) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L32)):*
`s/onehundred/oneHundred/g`
`s/onethousand/oneThousand/g`

Or maybe these should be `bigInt10`, `bigInt100`, `bigInt1000`.

---

*[util/encoding/decimal.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hYGjET03L_hKjHCh:-KD3hYGjET03L_hKjHCi:-1843782175) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L43)):*
Should we reserve space in the tag prefix for decimal NaN and infinity? Hard to add later if we don't reserve now.

---

*[util/encoding/decimal.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hvj6ZL8l7TwFx3A1:-KD3hvj7gUZqLq7fxScK:-1780580998) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L82)):*
`s/diving/stripping/g`?

---

*[util/encoding/decimal.go, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3iWyZw3CYtyxfs967:-KD3iWy_Sx6j0-tuEJlG:-1772406001) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L152)):*
`s/returnining/returning/g`

---

*[util/encoding/decimal.go, line 208 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3l5TAc2UmGFk0aJ9y:-KD3l5TAc2UmGFk0aJ9z:-1300286262) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L208)):*
Add a TODO about these cases being unimplemented.

---

*[util/encoding/decimal.go, line 299 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3mAMVoNCIx9I-e7Be:-KD3mAMVoNCIx9I-e7Bf:1795139877) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L299)):*
I wonder if this code predated `unsafe.Sizeof`. I think the above can be replaced with:

```
const _S = unsafe.Sizeof(big.Word(0))
```

And let's rename `_S` to something more comprehensible: `bigWordSize`.

---

*[util/encoding/decimal.go, line 336 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3gARxb4LadLcKxgS0:-KD3gARxb4LadLcKxgS1:-1036433949) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L336)):*
Since your index is in the range `[1,128]`, this would be faster as an array `[129]tableVal`.

---

*[util/encoding/encoding.go, line 739 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3gkJS3kibLBeYUfmH:-KD3gkJS3kibLBeYUfmI:2102611491) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/encoding.go#L739)):*
I think you should use the `'g'` format specifier. This would revert the changes to the pretty-printed output.

---

*[util/encoding/float.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3f_Nn02UzDSBCziad:-KD3f_Nn02UzDSBCziae:1448666868) ([raw file](https://github.com/cockroachdb/cockroach/blob/8df5343e2f07f040b93349ef050294a264d55db1/util/encoding/float.go#L52)):*
Do we need to special case 0 here? If we let this fall through below, the encoding of `-0` and `0` would be different which I think is desirable. Even if we do special case `0`, it might be better to do this below so that `-0` is encoded differently.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337)*

**Comment by nvanbenschoten** [comment]
Review status: 1 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[util/encoding/decimal.go, line 32 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hDSVBsOimZfBA5kd:-KD4fDOOdkV0Kml-PuOF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L32)):*
Done.

---

*[util/encoding/decimal.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hYGjET03L_hKjHCh:-KD4LQneCWX3bBWUKY1X:1007255830) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L43)):*
Do you mean create separate prefix tags for them? If so, they're already there. See the top of `encoding.go`.

---

*[util/encoding/decimal.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hvj6ZL8l7TwFx3A1:-KD4fKSixvGSToKPr6y2:-1952001725) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L82)):*
I think I was going for dividing.

---

*[util/encoding/decimal.go, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3iWyZw3CYtyxfs967:-KD4fSHfAgBgCtDMW9Tb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L152)):*
Done.

---

*[util/encoding/decimal.go, line 208 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3l5TAc2UmGFk0aJ9y:-KD4fdQQQVnGeLHJZZ96:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L208)):*
Done.

---

*[util/encoding/decimal.go, line 299 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3mAMVoNCIx9I-e7Be:-KD4ftUUPymwjrakJudk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L299)):*
Done.

---

*[util/encoding/decimal.go, line 336 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3gARxb4LadLcKxgS0:-KD59SJkCevX33HXYhiX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L336)):*
Done.

---

*[util/encoding/encoding.go, line 739 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3gkJS3kibLBeYUfmH:-KD5Ao4qelXexZIjW27v:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/encoding.go#L739)):*
Done.

---

*[util/encoding/float.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3f_Nn02UzDSBCziad:-KD5FDU6ZYQxDz--bsWp:2084035629) ([raw file](https://github.com/cockroachdb/cockroach/blob/8df5343e2f07f040b93349ef050294a264d55db1/util/encoding/float.go#L52)):*
Still special casing zero, but added support for negative zero.

---

*[util/encoding/float.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD2nZuJwji23jnGinsh-r2-55:-KD5HsMX965_p55tOEUH:944881988) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/float.go#L55)):*
Good point, what I had before was inverting the value a few too many times. The only thing that needed to be added to your example was that the negative high bit needs to be maintained when complementing the bits for negatives. Glad that we can now use `DecodeUint64Ascending` for both cases.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337)*

**Comment by petermattis** [comment]
:lgtm:

---

Reviewed 9 of 9 files at r2.
Review status: 3 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[util/encoding/decimal.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD3hYGjET03L_hKjHCh:-KD6-te2puZNxxgKnsi6:1050911931) ([raw file](https://github.com/cockroachdb/cockroach/blob/beab3d0f08c5b53a9facfc47b172c7de2140f382/util/encoding/decimal.go#L43)):*
Ah, good. Carry on.

---

*[util/encoding/float.go, line 50 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60A2ibis9wbrj02Cf:-KD60A2jGmHbGOJQemw6:-1660849949) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float.go#L50)):*
I think this can be `case u == 0`. See http://play.golang.org/p/oplF-iQu_l

---

*[util/encoding/float.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60QEBFqEU9LP0z9cv:-KD60QEBFqEU9LP0z9cw:-1743509331) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float.go#L56)):*
I don't think you need to or in the high bit again. Maybe I'm missing something. Can you provide some rationale? 

Hmm, I suppose adding the high-bit back allows you to keep the test data unchanged. Is that worth it? Maybe, I realize changing that test data is irritating. But seems better to bite-the-bullet and get rid of this.

---

*[util/encoding/float_test.go, line 88 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60o60Wpv5RqK-5vSB:-KD60o60Wpv5RqK-5vSC:610327414) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float_test.go#L88)):*
Remove?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-:-KD615fXc6XnF3PmBfLT:1417088721)*

**Comment by nvanbenschoten** [comment]
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/encoding/float.go, line 50 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60A2ibis9wbrj02Cf:-KD6QeSQYAk_TsuQs0h7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float.go#L50)):*
Done.

---

*[util/encoding/float.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60QEBFqEU9LP0z9cv:-KD6RYcz9EHYxfEICniD:334565594) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float.go#L56)):*
No you're completely right, since we're already appending a tag to distinguish positive and negative encoded floats, we don't need the upper bit to differentiate them as well. Removed and switched over the test data.

---

*[util/encoding/float_test.go, line 88 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-KD60o60Wpv5RqK-5vSB:-KD6QguNGg6oj3RlCvcp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73b77e191828535e57861c6e2eecd99008593e5e/util/encoding/float_test.go#L88)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337)*

**Comment by petermattis** [comment]
:shipit: 

---

Reviewed 7 of 7 files at r3.
Review status: 6 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5337#-:-KD8qeiVyAnL29q4QX6f:1148172450)*

**Pull request merged**